File: v1/src/main/java/com/api/v1/medical_slots/services/MedicalSlotRetrievalServiceImpl.java
package com.api.v1.medical_slots.services;

import com.api.v1.common.LicenseNumber;
import com.api.v1.common.ObjectId;
import com.api.v1.doctors.Doctor;
import com.api.v1.doctors.DoctorFinder;
import com.api.v1.medical_slots.MedicalSlot;
import com.api.v1.medical_slots.MedicalSlotFinder;
import com.api.v1.medical_slots.MedicalSlotResponseDTO;
import com.api.v1.medical_slots.controllers.MedicalSlotController;
import com.api.v1.medical_slots.domain.MedicalSlotCrudRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;

@Service
public class MedicalSlotRetrievalServiceImpl implements MedicalSlotRetrievalService {

    private final MedicalSlotCrudRepository crudRepository;
    private final DoctorFinder doctorFinder;
    private final MedicalSlotFinder medicalSlotFinder;

    public MedicalSlotRetrievalServiceImpl(MedicalSlotCrudRepository crudRepository,
                                           DoctorFinder doctorFinder,
                                           MedicalSlotFinder medicalSlotFinder
    ) {
        this.crudRepository = crudRepository;
        this.doctorFinder = doctorFinder;
        this.medicalSlotFinder = medicalSlotFinder;
    }

    @Override
    public ResponseEntity<MedicalSlotResponseDTO> findByIdAndDoctor(@ObjectId String id,
                                                                    @LicenseNumber String medicalLicenseNumber
    ) {
        Doctor foundDoctor = doctorFinder.findByLicenseNumber(medicalLicenseNumber);
        MedicalSlot foundSlot = medicalSlotFinder.findByIdAndDoctor(id, foundDoctor);
        MedicalSlotResponseDTO responseDTO = foundSlot.toDTO();
        responseDTO.add(
            linkTo(methodOn(MedicalSlotController.class).findByIdAndDoctor(id, medicalLicenseNumber)).withSelfRel()
        );
        return ResponseEntity.ok(responseDTO);
    }

    @Override
    public ResponseEntity<Page<MedicalSlotResponseDTO>> findAllByDoctor(@LicenseNumber String medicalLicenseNumber,
                                                                        Pageable pageable
    ) {
        Doctor foundDoctor = doctorFinder.findByLicenseNumber(medicalLicenseNumber);
        var all = crudRepository.findAllByDoctor(foundDoctor.getId(), pageable).map(MedicalSlot::toDTO);
        return ResponseEntity.ok(all);
    }

    @Override
    public ResponseEntity<Page<MedicalSlotResponseDTO>> findAll(Pageable pageable) {
        var all = crudRepository.findAll(pageable).map(MedicalSlot::toDTO);
        return ResponseEntity.ok(all);
    }
}

File: v2/src/main/kotlin/com/api/v2/medical_slots/services/MedicalSlotRetrievalServiceImpl.kt
package com.api.v2.medical_slots.services

import com.api.v2.doctors.DoctorFinder
import com.api.v2.medical_slots.MedicalSlot
import com.api.v2.medical_slots.MedicalSlotFinder
import com.api.v2.medical_slots.MedicalSlotResponseDTO
import com.api.v2.medical_slots.controllers.MedicalSlotController
import com.api.v2.medical_slots.domain.MedicalSlotCrudRepository
import com.api.v2.medical_slots.toDTO
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service

@Service
class MedicalSlotRetrievalServiceImpl(
    private val crudRepository: MedicalSlotCrudRepository,
    private val doctorFinder: DoctorFinder,
    private val medicalSlotFinder: MedicalSlotFinder
): MedicalSlotRetrievalService {

    override fun findByIdAndDoctor(
        id: String,
        medicalLicenseNumber: String
    ): ResponseEntity<MedicalSlotResponseDTO> {
        val foundDoctor = doctorFinder.findByLicenseNumber(medicalLicenseNumber)
        val foundSlot = medicalSlotFinder.findByIdAndDoctor(id, foundDoctor)
        val dto = foundSlot.toDTO()
        dto.add(
            linkTo(methodOn(MedicalSlotController::class.java)
                .findByIdAndDoctor(id, medicalLicenseNumber)
            ).withSelfRel()
        )
        return ResponseEntity.ok(dto)
    }

    override fun findAllByDoctor(
        medicalLicenseNumber: String,
        pageable: Pageable
    ): ResponseEntity<Page<MedicalSlotResponseDTO>> {
        val foundDoctor = doctorFinder.findByLicenseNumber(medicalLicenseNumber)
        val page = crudRepository.findAllByDoctor(foundDoctor.id, pageable).map(MedicalSlot::toDTO)
        return ResponseEntity.ok(page)
    }

    override fun findAll(pageable: Pageable): ResponseEntity<Page<MedicalSlotResponseDTO>> {
        val page = crudRepository.findAll(pageable).map(MedicalSlot::toDTO)
        return ResponseEntity.ok(page)
    }
}